File: src/data/finnhub.ts
const BASE = "https://finnhub.io/api/v1"

export interface FinnhubQuote {
  price: number
  currency: string
  change: number
  changePercent: number
}

export interface FinnhubNewsItem {
  title: string
  url: string
  publishedAt: string
  source: string
}

interface QuoteResponse {
  c?: number // current
  d?: number // change
  dp?: number // percent change
}

interface NewsResponse {
  headline?: string
  url?: string
  datetime?: number // unix seconds
  source?: string
}

function key(): string | undefined {
  return process.env.FINNHUB_API_KEY
}

export async function fetchFinnhubQuote(symbol: string): Promise<FinnhubQuote | null> {
  const token = key()
  if (!token) return null
  const res = await fetch(`${BASE}/quote?symbol=${encodeURIComponent(symbol)}&token=${token}`)
  if (!res.ok) return null
  const data = (await res.json()) as QuoteResponse
  if (data.c == null || data.c === 0) return null
  return {
    price: data.c,
    currency: "USD",
    change: data.d ?? 0,
    changePercent: data.dp ?? 0,
  }
}

export async function fetchFinnhubNews(symbol: string): Promise<FinnhubNewsItem[]> {
  const token = key()
  if (!token) return []
  const to = new Date()
  const from = new Date(to.getTime() - 14 * 24 * 60 * 60 * 1000)
  const fmt = (d: Date): string => d.toISOString().slice(0, 10)
  const url = `${BASE}/company-news?symbol=${encodeURIComponent(symbol)}&from=${fmt(from)}&to=${fmt(to)}&token=${token}`
  const res = await fetch(url)
  if (!res.ok) return []
  const data = (await res.json()) as NewsResponse[]
  if (!Array.isArray(data)) return []
  return data
    .filter((n) => n.headline && n.url)
    .map((n) => ({
      title: n.headline as string,
      url: n.url as string,
      publishedAt: n.datetime ? new Date(n.datetime * 1000).toISOString() : new Date().toISOString(),
      source: n.source ?? "Finnhub",
    }))
}

File: app/data/finnhub.py
"""Finnhub REST access (optional; real-time US quotes + company news)."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta

import httpx

from ..config import FINNHUB_API_KEY
from ..schemas import NewsItem, Quote

BASE = "https://finnhub.io/api/v1"


async def fetch_quote(symbol: str) -> Quote | None:
    if not FINNHUB_API_KEY:
        return None
    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.get(
            f"{BASE}/quote", params={"symbol": symbol, "token": FINNHUB_API_KEY}
        )
    if res.status_code != 200:
        return None
    data = res.json()
    current = data.get("c")
    if not current:  # None or 0 -> no usable quote
        return None
    return Quote(
        price=float(current),
        currency="USD",
        change=float(data.get("d") or 0.0),
        change_percent=float(data.get("dp") or 0.0),
    )


async def fetch_news(symbol: str) -> list[NewsItem]:
    if not FINNHUB_API_KEY:
        return []
    to = datetime.now(UTC)
    frm = to - timedelta(days=14)
    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.get(
            f"{BASE}/company-news",
            params={
                "symbol": symbol,
                "from": frm.date().isoformat(),
                "to": to.date().isoformat(),
                "token": FINNHUB_API_KEY,
            },
        )
    if res.status_code != 200:
        return []
    data = res.json()
    if not isinstance(data, list):
        return []
    out: list[NewsItem] = []
    for n in data:
        headline, url = n.get("headline"), n.get("url")
        if not headline or not url:
            continue
        ts = n.get("datetime")
        published = (
            datetime.fromtimestamp(ts, tz=UTC).isoformat()
            if ts
            else datetime.now(UTC).isoformat()
        )
        out.append(
            NewsItem(
                title=headline, url=url, published_at=published, source=n.get("source") or "Finnhub"
            )
        )
    return out
